### ISSUE TITLE
500 Error when posting update to a page

### ISSUE THREAD

**Issue by farktronix** [rpj/spring83#12]
When I try to post an update to AMERICA'S HOTTEST TALKLINE I'm getting a 500 internal server error:

```
./putnew https://0l0.lol c7f42a28888a80322418faa5d65b4410847bdfca0e220a415689702d661424c0 myboards/6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.html
6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.html: appended tags added 44 bytes
ERROR: POST https://0l0.lol 500 "Internal Server Error"
node:internal/errors:477
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received undefined
    at new NodeError (node:internal/errors:388:5)
    at Function.from (node:buffer:326:9)
    at preprocess (/Users/pharkas/code/spring83/putnew:63:28)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async putOne (/Users/pharkas/code/spring83/putnew:105:25)
    at async main (/Users/pharkas/code/spring83/putnew:139:3) {
  code: 'ERR_INVALID_ARG_TYPE'
}

Node.js v18.4.0
```

Here's the HTML (Github won't let me upload it as .html so it's renamed to .txt): 
[6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.txt](https://github.com/rpj/spring83/files/9006195/6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.txt)

```
$ putnew --version
1.2.0
```

Note: I don't have time to debug this right now so I'm leaving this for myself to follow up on later (or feel free to fix this if you know what it is and it's easy!)

**Comment by rpj** [comment]
Well shoot, that's no good! Unfortunately I don't have time to debug now myself either _but_ you can get around that entirely (avoiding that entire code block) by passing `--no-shorten`: it won't shorten links but it will still minify (unless you've also included `--no-minify` of course).

**Comment by rpj** [comment]
Did manage to log in and grab the back trace before it gets annoyingly buried in logs:

```
TypeError: Cannot read properties of undefined (reading startsWith)
    at Element.<anonymous> (/usr/src/spring83/server/routes.js:126:32)
    at LoadedCheerio.each (/usr/src/spring83/node_modules/cheerio/lib/api/traversing.js:519:26)
    at /usr/src/spring83/server/routes.js:125:21
    at postPutHandlerValidator (/usr/src/spring83/server/routes.js:113:12)
```

**Comment by rpj** [comment]
Argh, this one is an infuriating dumb mistake on my part, and a literal one-character fix. 🤦  Not all `a` tags must have a `href` attribute!

```diff
diff --git a/server/routes.js b/server/routes.js
index 736c1c3..0d9c549 100644
--- a/server/routes.js
+++ b/server/routes.js
@@ -123,7 +123,7 @@ async function attach (app, knownKeys, fqdn, contentDir, contactAddr, scheme) {
         const ignoreBoardLinks = disabled.includes('shorten-board-links');
         const promises = [];
         parsed('a').each((idx, ele) => {
-          if (ele.attribs.href.startsWith('http')) {
+          if (ele.attribs.href?.startsWith('http')) {
             promises.push(async () => {
               const { host, pathname } = new URL(ele.attribs.href);
               if (constants.shortener.knownS83Hosts.includes(host) && pathname.indexOf('/') === 0) {
```

and sure enough, that fixes it (100% reproducible with the HTML you attached 🙌. Almost like you had done world-class QA engineering at some point or something... 😝 ❤️ )

## before fix:
```shell
$ ~/repos/spring83/putnew --doNotShare true --postHost http://0.0.0.0:1783 http://0.0.0.0:1783 c7f42a28888a80322418faa5d65b4410847bdfca0e220a415689702d661424c0 issue-12.html
6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.html: appended tags added 86 bytes
ERROR: POST http://0.0.0.0:1783 500 "Internal Server Error"
node:internal/errors:466
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received undefined
    at new NodeError (node:internal/errors:377:5)
    at Function.from (node:buffer:325:9)
    at preprocess (C:\Users\ryanj\repos\spring83\putnew:63:28)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async putOne (C:\Users\ryanj\repos\spring83\putnew:105:25)
    at async main (C:\Users\ryanj\repos\spring83\putnew:139:3) {
  code: 'ERR_INVALID_ARG_TYPE'
}

Node.js v18.2.0
```
## with fix:
```shell
$ ~/repos/spring83/putnew --doNotShare true --postHost http://0.0.0.0:1783 http://0.0.0.0:1783 c7f42a28888a80322418faa5d65b4410847bdfca0e220a415689702d661424c0 issue-12.html
6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.html: appended tags added 86 bytes
POST http://0.0.0.0:1783 200
6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323.html: minified 25% (574 bytes) -> 1742 total
PUT http://0.0.0.0:1783 200 -> http://0.0.0.0:1783/6450bc6ba37ccf5390c3d46034bbcf518d31cee1cda5c281cf5c19df283e0323
```

Pushed fixed in (`c962e29b`)[https://github.com/rpj/spring83/commit/c962e29b9737b613c39203fc72b2576c5f94dd26]. Sorry to steal the fix from you but I felt badly that this was broken for you. 

Also, Chase is already experimenting with a "chat room board" apparently too, so you need to get yours on the board directory to show folks that _we're already doing it_ 😆!  It is just controlled directly by Robins so shoot him an email editor@followersentinel.com to have it added
